libs/shared/src/main/java/io/crate/common/collections/accountable/AccountableList.java

```diff
@@ -0,0 +1,180 @@
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.common.collections.accountable;
+
+import java.util.AbstractList;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.RandomAccess;
+import java.util.function.IntConsumer;
+
+/**
+ * List implementation accounting for shallow memory usage.
+ * Based on {@link ArrayList} with some simplifications.
+ * Doesn't have concurrent modification check.
+ * Supports boundary checks since some features (lag/lead) rely on IndexOutOfBoundsException.
+ * Supports only {@link #add(Object)} operation.
+ */
+public class AccountableList<T> extends AbstractList<T> {
+
+    private static final int SOFT_MAX_ARRAY_LENGTH = Integer.MAX_VALUE - 8;
+    private static final Object[] DEFAULTCAPACITY_EMPTY_ELEMENTDATA = {};
+    private static final int DEFAULT_CAPACITY = 10;
+
+    private final IntConsumer growthAccounting;
+    private Object[] elementData;
+    private int size;
+
+    public AccountableList(IntConsumer growthAccounting) {
+        this.growthAccounting = growthAccounting;
+        this.elementData = DEFAULTCAPACITY_EMPTY_ELEMENTDATA;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public T get(int index) {
+        Objects.checkIndex(index, size);
+        return (T) elementData[index];
+    }
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    /**
+     * Similar to {@link ArrayList#sort(Comparator)} but without modification count checks.
+     */
+    public void sort(Comparator<? super T> c) {
+        Arrays.sort((T[]) elementData, 0, size, c);
+    }
+
+    @Override
+    public boolean add(T t) {
+        if (size == elementData.length) {
+            elementData = grow(size + 1);
+        }
+        elementData[size] = t;
+        size++;
+        return true;
+    }
+
+    /**
+     *
+     * @param fromIndex low endpoint (inclusive) of the subList
+     * @param toIndex high endpoint (exclusive) of the subList
+     * @return
+     */
+    public List<T> subList(int fromIndex, int toIndex) {
+        return new SubList<>(this, fromIndex, toIndex);
+    }
+
+    private static class SubList<T> extends AbstractList<T> implements RandomAccess {
+        private final AccountableList<T> root;
+        private final int offset;
+        private int size;
+
+        /**
+         * Constructs a sublist of an arbitrary ArrayList.
+         */
+        public SubList(AccountableList<T> root, int fromIndex, int toIndex) {
+            this.root = root;
+            this.offset = fromIndex;
+            this.size = toIndex - fromIndex;
+        }
```

Comment:
We can even account for an object header. Need to re-use `addBytes` instead of using `growthAccouting`. 
See https://github.com/crate/crate/pull/16263#discussion_r1665233994